sdk/core/azure-core/azure/core/utils/_utils.py

```diff
@@ -87,26 +87,79 @@ def case_insensitive_dict(*args: Any, **kwargs: Any) -> MutableMapping:
     :return: A case-insensitive mutable mapping object.
     :rtype: ~collections.abc.MutableMapping
     """
+    return CaseInsensitiveDict(*args, **kwargs)
+
+class CaseInsensitiveDict(MutableMapping):
+    """
+    NOTE: This implementation is heavily inspired from the case insensitive dictionary from the requests library. Thank you !!
+    Case insensitive dictionary implementation.
+    The keys are expected to be strings and will be stored in lower case.
+    case_insensitive_dict = CaseInsensitiveDict()
+    case_insensitive_dict['Key'] = 'some_value'
+    case_insensitive_dict['key'] == 'some_value' #True
+    """
+    def __init__(self, data = None, **kwargs: Any) -> None:
+        self._store = {}
+        if data is None:
+            data = {}
+        
+        self.update(data, **kwargs)
+
+
+    def copy(self) -> "CaseInsensitiveDict":
+        return CaseInsensitiveDict(self._store.values())
+
+    def __getitem__(self, key: str) -> Any:
+        return self._store[key.lower()]
+
+    def __setitem__(self, key: str, value: str) -> None:
+        """
+        Set the `key` to `value`. The original key will be stored with the value
+        """
+        self._store[key.lower()] = (key, value)
+
+    def __getitem__(self, key: str) -> Any:
+        return self._store[key.lower()][1]
+
+    def __delitem__(self, key: str) -> None:
+        del self._store[key.lower()]
+
+    def __iter__(self) -> Iterable[str]:
+        return (key for key, _ in self._store.values())
+
+    def __len__(self) -> int:
+        return len(self._store)
+
+    def lowerkey_items(self):
```

Comment:
is this something we want exposed?

Reply:
I had initially kept it private but then it came in to use when I implemented `__eq__` as I would need to access the other dict items, so didn't want to access a private method. It was mostly preference to keep it public because of equals